Libs/MRML/Core/vtkMRMLSegmentationDisplayNode.cxx

```diff
@@ -365,62 +368,121 @@ void vtkMRMLSegmentationDisplayNode::SetSegmentDisplayProperties(std::string seg
       }
     }
 
-  // Save cached value of color
-  // TODO: remove this when terminology infrastructure is in place
+  if (modified)
+    {
+    this->Modified();
+    }
+}
+
+//---------------------------------------------------------------------------
+vtkVector3d vtkMRMLSegmentationDisplayNode::GetSegmentColor(std::string segmentID)
+{
+  double color[3] = {vtkSegment::SEGMENT_COLOR_INVALID[0], vtkSegment::SEGMENT_COLOR_INVALID[1], vtkSegment::SEGMENT_COLOR_INVALID[2]};
+  this->GetSegmentColor(segmentID, color);
+
+  vtkVector3d colorVtk(color[0], color[1], color[2]);
+  return colorVtk;
+}
+
+//---------------------------------------------------------------------------
+void vtkMRMLSegmentationDisplayNode::GetSegmentColor(std::string segmentID, double* color)
+{
+  if (!color)
+    {
+    vtkErrorMacro("GetSegmentColor: Invalid output color array");
+    return;
+    }
+
+  // Invalidate color
+  color[0] = vtkSegment::SEGMENT_COLOR_INVALID[0];
+  color[1] = vtkSegment::SEGMENT_COLOR_INVALID[1];
+  color[2] = vtkSegment::SEGMENT_COLOR_INVALID[2];
+
+  // Get segment
   vtkMRMLSegmentationNode* segmentationNode = vtkMRMLSegmentationNode::SafeDownCast(this->GetDisplayableNode());
-  if (segmentationNode)
+  if (!segmentationNode)
     {
-    vtkSegmentation* segmentation = segmentationNode->GetSegmentation();
-    vtkSegment* segment = segmentation->GetSegment(segmentId);
-    if (segment)
-      {
-      segment->SetDefaultColorWithoutModifiedEvent(properties.Color);
-      }
+    vtkErrorMacro("GetSegmentColor: No segmentation node associated to this display node");
+    return;
+    }
+  vtkSegmentation* segmentation = segmentationNode->GetSegmentation();
+  vtkSegment* segment = segmentation->GetSegment(segmentID);
+  if (!segment)
+    {
+    vtkErrorMacro("GetSegmentColor: segment not found by id " << segmentID);
+    return;
     }
 
-  if (modified)
+  // Get color from segment
+  segment->GetColor(color);
+
+  // Get display properties
+  this->UpdateSegmentList();
+  SegmentDisplayPropertiesMap::iterator propsIt = this->SegmentationDisplayProperties.find(segmentID);
+  if ( propsIt != this->SegmentationDisplayProperties.end()
+    && propsIt->second.OverrideColor[0] != SEGMENT_COLOR_NO_OVERRIDE
```

Comment:
It might be safer to check (overrideColor[0]>=0 && overrideColor[1]>=0 && overrideColor[2]>=0) to make sure the color is fully valid